autogpt_platform/backend/backend/blocks/codex.py

```diff
@@ -0,0 +1,280 @@
+            advanced=True,
+        )
+        credentials: CodexCredentials = CodexCredentialsField()
+
+    class Output(BlockSchemaOutput):
+        response: str = SchemaField(
+            description="Code-focused response returned by the Codex model."
+        )
+        reasoning: str = SchemaField(
+            description="Reasoning summary returned by the model, if available.",
+            default="",
+        )
+        response_id: str = SchemaField(
+            description="ID of the Responses API call for auditing/debugging.",
+            default="",
+        )
+
+    def __init__(self):
+        super().__init__(
+            id="86a2a099-30df-47b4-b7e4-34ae5f83e0d5",
+            description="Generate or refactor code using OpenAI's Codex (Responses API).",
+            categories={BlockCategory.AI, BlockCategory.DEVELOPER_TOOLS},
+            input_schema=CodexBlock.Input,
+            output_schema=CodexBlock.Output,
+            test_input=[
+                {
+                    "prompt": "Write a TypeScript function that deduplicates an array.",
+                    "credentials": TEST_CREDENTIALS_INPUT,
+                }
+            ],
+            test_output=[
+                ("response", str),
+                ("reasoning", str),
+                ("response_id", str),
+            ],
+            test_mock={
+                "call_codex": lambda *args, **kwargs: {
+                    "response": "function dedupe<T>(items: T[]): T[] { return [...new Set(items)]; }",
+                    "reasoning": "Used Set to remove duplicates in O(n).",
+                    "response_id": "resp_test",
+                }
+            },
+            test_credentials=TEST_CREDENTIALS,
+        )
+        self.execution_stats = NodeExecutionStats()
+
+    async def call_codex(
+        self,
+        *,
+        credentials: APIKeyCredentials,
+        model: CodexModel,
+        prompt: str,
+        system_prompt: str,
+        max_output_tokens: int | None,
+        reasoning_effort: CodexReasoningEffort,
+    ) -> dict[str, Any]:
+        """Invoke the OpenAI Responses API."""
+        client = AsyncOpenAI(api_key=credentials.api_key.get_secret_value())
+
+        request_payload: dict[str, Any] = {
+            "model": model.value,
+            "input": prompt,
+        }
+        if system_prompt:
+            request_payload["instructions"] = system_prompt
+        if max_output_tokens is not None:
+            request_payload["max_output_tokens"] = max_output_tokens
+        if reasoning_effort != CodexReasoningEffort.NONE:
+            request_payload["reasoning"] = {"effort": reasoning_effort.value}
+
+        response = await client.responses.create(**request_payload)
+
+        text_output = self._extract_output_text(response)
+        reasoning_summary = self._extract_reasoning_summary(response)
+        response_id = getattr(response, "id", "")
```

Comment:
with actual typing you will no longer need all of these duck-typing like getattr isinstance etc